### PR TITLE
Optimize search write position in BlobFile

### LINKED ISSUE #6404: searchInsertOffset takes too much time cause Region xxx is unavailable error
The write process is slow due to `searchInsertOffset` taking too much time. It is an O(n) operation.

image

image

It consumes over 60% of the time of applying raft logs, with 963k pages.

### PR DESCRIPTION
**Pull request by lidezhu**
### What problem does this PR solve?

Issue Number: close #6404 

Problem Summary: The process to search an available space in BlobFile is an O(n) operation and introduce too much latency for the whole write process.

### What is changed and how it works?
1. Introduce a `free_map_invert_index` to eliminate the O(n) operation;
2. Skip BlobFiles which are already locked when search for an available space for write;

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/Page/V3/spacemap/SpaceMapSTDMap.h [hunk 1]
```diff
@@ -55,6 +55,8 @@ class STDMapSpaceMap
         : SpaceMap(start, end, SMAP64_STD_MAP)
     {
         free_map.insert({start, end});
+        std::set<UInt64> offsets{start};
+        free_map_invert_index.emplace(end, offsets);
```
Comment (flowbehappy) [inline dbms/src/Storages/Page/V3/spacemap/SpaceMapSTDMap.h:59]: `end` is not exactly equal to "size"? Maybe `end - start` is more accurate?

File: dbms/src/Storages/Page/V3/spacemap/SpaceMapSTDMap.h [hunk 2]
```diff
@@ -416,30 +331,49 @@ class STDMapSpaceMap
...
             free_map.erase(it_next);
         }
         // next can't merge
         return true;
     }
 
+private:
+    inline void deleteFromInvertIndex(UInt64 length, UInt64 offset)
+    {
+        auto & offsets = free_map_invert_index[length];
```
Comment (flowbehappy) [inline dbms/src/Storages/Page/V3/spacemap/SpaceMapSTDMap.h:347]: How about doing some checking here?  Like print some message if `free_map_invert_index[length]` does not exist.

Comment (lidezhu) [reply dbms/src/Storages/Page/V3/spacemap/SpaceMapSTDMap.h]: Done.

File: dbms/src/Storages/Page/V3/spacemap/SpaceMapSTDMap.h (hunk 1 above)
Comment (lidezhu) [reply dbms/src/Storages/Page/V3/spacemap/SpaceMapSTDMap.h]: Fixed.

File: dbms/src/Common/TiFlashMetrics.h [hunk 3]
```diff
@@ -178,10 +178,13 @@ namespace DB
     M(tiflash_storage_page_write_duration_seconds, "The duration of each write batch", Histogram,                                         \
         F(type_total, {{"type", "total"}}, ExpBuckets{0.0001, 2, 20}),                                                                    \
         F(type_blob,  {{"type", "blob"}},  ExpBuckets{0.0001, 2, 20}),                                                                    \
```
Comment (JaySon-Huang) [inline dbms/src/Common/TiFlashMetrics.h:180]: We can remove this since we break it down to `choose_stat`/`search_pos`/`blob_write` in this PR

**Review by JaySon-Huang — APPROVED** [review]
Better have a micro benchmark of searching insert offset in a space map [0, 512MB) with many small blocks less than 4KB

else LGTM

File: dbms/src/Common/TiFlashMetrics.h [hunk 4]
```diff
@@ -176,12 +176,14 @@ namespace DB
     M(tiflash_storage_page_write_batch_size, "The size of each write batch in bytes", Histogram,                                          \
         F(type_v3, {{"type", "v3"}}, ExpBuckets{4 * 1024, 4, 10}))                                                                        \
     M(tiflash_storage_page_write_duration_seconds, "The duration of each write batch", Histogram,                                         \
-        F(type_total, {{"type", "total"}}, ExpBuckets{0.0001, 2, 20}),                                                                    \
-        F(type_blob,  {{"type", "blob"}},  ExpBuckets{0.0001, 2, 20}),                                                                    \
+        F(type_total,       {{"type", "total"}},       ExpBuckets{0.0001, 2, 20}),                                                        \
+        F(type_choose_stat, {{"type", "choose_stat"}}, ExpBuckets{0.0001, 2, 20}),                                                        \
```
[the remark this answers is not in the document]
Comment (lidezhu) [reply dbms/src/Common/TiFlashMetrics.h]: Done.

**Comment by lidezhu** [comment]
> Better have a micro benchmark of searching insert offset in a space map [0, 512MB) with many small blocks less than 4KB
> 
> else LGTM

OK. I will add it.

**Review by flowbehappy — APPROVED** [review]
LGTM

**Pull request merged**